### PR TITLE
Module4

### PR DESCRIPTION
**Pull request by dzmitrymatus**
Module4 implementation

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -4,26 +4,33 @@
     // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
     "version": "0.2.0",
     "configurations": [
+        {
```
Comment (av-shevchenko89) [inline .vscode/launch.json:7]: please add it to .gitignore

File: courses-app/src/app/features/registration/registration.component.ts [hunk 2]
```diff
@@ -7,9 +12,38 @@ import { Component, OnInit } from '@angular/core';
 })
 export class RegistrationComponent implements OnInit {
 
-  constructor() { }
+  registrationForm : FormGroup;
+
+  constructor(private authService: AuthService,
+    private router: Router) {
+    this.registrationForm = new FormGroup({
+      "name" : new FormControl("", [Validators.required, Validators.minLength(6)]),
```
Comment (av-shevchenko89) [inline courses-app/src/app/features/registration/registration.component.ts:20]: "" is not mandatory

File: courses-app/src/app/services/authors/authors-store.service.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  private authors$$: BehaviorSubject<AuthorModel[]> = new BehaviorSubject<AuthorModel[]>([]);
+  private isLoading$$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+
+  public authors$: Observable<AuthorModel[]> = this.authors$$.asObservable();
+  public isLoading$: Observable<boolean> = this.isLoading$$.asObservable();
+
+  constructor(private authorsService: AuthorsService) { }
+
+  getAll(): Observable<AuthorModel[]> {
+    return defer(() => {
```
Comment (av-shevchenko89) [inline courses-app/src/app/services/authors/authors-store.service.ts:20]: 'return defer(() => {' is not necessary

File: courses-app/src/app/services/courses/courses-store.service.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        return newItem;
+      })
+    ));
+
+  constructor(private coursesService: CoursesService) {
+    this.coursesFromServer$$.subscribe(data => this.courses$$.next(data));
+   }
+
+  getAll(): Observable<CourseModel[]> {
+    return defer(() => {
```
Comment (av-shevchenko89) [inline courses-app/src/app/services/courses/courses-store.service.ts:29]: 'return defer(() => {' is not necessary

File: courses-app/src/app/shared/components/button/button.component.ts [hunk 5]
```diff
@@ -22,17 +23,21 @@ export class ButtonComponent implements OnInit {
...
-   onButtonClick() {
-      if(this.confirmWindow !== undefined){
-        this.confirmWindow.show();
-        this.confirmWindow.onClose.subscribe(x => 
-          {
-            if(x === true) this.onClick.emit();
-            this.confirmWindow?.hide();
-          });
-      }
+    if(this.confirmWindow !== undefined){
```
Comment (av-shevchenko89) [inline courses-app/src/app/shared/components/button/button.component.ts:27]: if (this.confirmWindow) {

File: courses-app/src/app/shared/components/button/button.component.ts [hunk 6]
```diff
@@ -22,17 +23,21 @@ export class ButtonComponent implements OnInit {
...
+      this.confirmWindow.onClose.subscribe(x => 
+        {
+          if(x === true) this.onClick.emit();
+          this.confirmWindow?.hide();
+        });
+    } 
    }
 
+  onButtonClick() {
+    if(this.confirmWindow === undefined) {
```
Comment (av-shevchenko89) [inline courses-app/src/app/shared/components/button/button.component.ts:37]: if (this.confirmWindow) {

File: courses-app/src/app/shared/pipes/creation-date/creation-date.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (av-shevchenko89) [inline courses-app/src/app/shared/pipes/creation-date/creation-date.pipe.ts:1]: do you really need this pipe? why not just use build-in

File: courses-app/src/environments/api.config.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+const baseUrl = "http://localhost:3000";
+
+export const apiConfig = {
+    baseUrl: baseUrl,
+    getAuthorsUrl: baseUrl + "/authors/all",
```
Comment (av-shevchenko89) [inline courses-app/src/environments/api.config.ts:5]: imho, this is not the best way to set env config on FE. there is a way to proxy to remote env with angular-cli. Also apiConfig in this case can be just - `getAuthorsUrl: "/authors/all",`

**Review by av-shevchenko89 — APPROVED** [review]

File: .vscode/launch.json (hunk 1 above)
Comment (dzmitrymatus) [reply .vscode/launch.json:7]: fixed in the next module

File: courses-app/src/app/features/registration/registration.component.ts (hunk 2 above)
Comment (dzmitrymatus) [reply courses-app/src/app/features/registration/registration.component.ts:20]: fixed in the next module

File: courses-app/src/app/shared/components/button/button.component.ts (hunk 5 above)
Comment (dzmitrymatus) [reply courses-app/src/app/shared/components/button/button.component.ts:27]: fixed in the next module

File: courses-app/src/app/shared/components/button/button.component.ts (hunk 6 above)
Comment (dzmitrymatus) [reply courses-app/src/app/shared/components/button/button.component.ts:37]: fixed in the next module

File: courses-app/src/app/shared/pipes/creation-date/creation-date.pipe.ts (hunk 7 above)
Comment (dzmitrymatus) [reply courses-app/src/app/shared/pipes/creation-date/creation-date.pipe.ts:1]: it was a part of task (module3)

File: courses-app/src/environments/api.config.ts (hunk 8 above)
Comment (dzmitrymatus) [reply courses-app/src/environments/api.config.ts:5]: fixed in the next module

File: courses-app/src/app/services/courses/courses-store.service.ts (hunk 4 above)
Comment (dzmitrymatus) [reply courses-app/src/app/services/courses/courses-store.service.ts:29]: got it, fixed in the next module

File: courses-app/src/app/services/authors/authors-store.service.ts (hunk 3 above)
Comment (dzmitrymatus) [reply courses-app/src/app/services/authors/authors-store.service.ts:20]: got it, fixed in the next module

**Pull request merged**